Core/SqlStatementModel/ConstantCollectionExpression.cs

```diff
@@ -0,0 +1,109 @@
+﻿// This file is part of the re-linq project (relinq.codeplex.com)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// re-linq is free software; you can redistribute it and/or modify it under
+// the terms of the GNU Lesser General Public License as published by the
+// Free Software Foundation; either version 2.1 of the License,
+// or (at your option) any later version.
+//
+// re-linq is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+//
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel
+{
+  /// <summary>
+  /// Expression representing an <see cref="ICollection"/>, an <see cref="ICollection{T}"/> or an <see cref="IReadOnlyCollection{T}"/>.
+  /// </summary>
+  public class ConstantCollectionExpression : Expression
+  {
+    /// <summary>
+    /// The original expression.
+    /// </summary>
+    public ConstantExpression ConstantExpression { get; }
+
+    /// <summary>
+    /// Indicates whether the <see cref="ConstantCollectionExpression.ConstantExpression"/>'s
+    /// <see cref="System.Linq.Expressions.ConstantExpression.Value"/> is a collection without elements.
+    /// </summary>
+    public bool IsEmptyCollection { get; }
+    
+    /// <summary>
+    /// The <see cref="System.Type"/> of the <see cref="ConstantCollectionExpression.ConstantExpression"/>'s <see cref="ConstantExpression.Value"/>.
+    /// </summary>
+    public override Type Type { get; }
+
+    /// <summary>
+    /// Creates a <see cref="ConstantCollectionExpression"/> if the given <paramref name="constantExpression"/> is a collection.
+    /// </summary>
+    /// <exception cref="ArgumentException">
+    ///   If <paramref name="constantExpression"/> is <see langword="null"/>,
+    ///   its <see cref="ConstantExpression.Value"/> is <see langword="null"/>
+    ///   or the <see cref="ConstantExpression.Value"/> is not a collection.
+    /// </exception>
+    public ConstantCollectionExpression (ConstantExpression constantExpression)
+    {
+      ArgumentUtility.CheckNotNull ("constantExpression", constantExpression);
+
+      var value = constantExpression.Value;
+      if (!CollectionUtility.IsCollection (value))
+        throw new ArgumentException ("The given constant expression does not have a collection value.", nameof(constantExpression));
+
+      ConstantExpression = constantExpression;
+      Type = value.GetType();
+
+      var collection = (IEnumerable) ConstantExpression.Value;
+      IsEmptyCollection = !collection.Cast<object>().Any();
```

Comment:
should we look into allocation free by first trying to cast to ICollection and IReadOnlyCollection<object> and only iterate if they don't succeed?

Reply:
Since this is only done once per `ConstantExpression` that contains a collection in its `Value`, I don't think that this optimization would be worth the additional code.

Reply:
This might be true in ExpressionTree analysis. It just feels wrong on a deeply personal level ;) I probably should stop reading how Ayende builds .NET libraries that won't allocate a single byte (said with a smidge of hyperbole)